Add OIDC client access token expires in skew

sberyozkin opened the pull request:
Fixes #46142.

This simple PR adds an OIDC client access token expires in property to let users get the correct token expiry time.

By default, the current time is added to the relative expires_in response property - however, due to the clock skew or the IDP specific configuration (as explained in the recent Zulip dev thread), the actual token issued at time can be earlier (or even later) than the current time. For example, a 1 min difference can be quite significant.

Adding an option to sync it with by configuring this new skew property (Duration can also be negative if needed) will let OIDC client calculate a more precise token expiry time. 

I've confirmed that updated test is failing without adding a skew property.

If this PR does not make it to 3.18 right now then we might Guillaume to consider it for the backport to 3.18 later, we don't have to rush  it if there are some concerns

sberyozkin committed:
Add OIDC client access token expires in skew

sberyozkin commented:
Michal, George approved so I added a `waiting for ci` label for Guillaume @gsmet or someone else to possibly merge tomorrow morning, I just wanted to keep you in the loop re the way the OIDC client configuration was updated (you made it very easy to add it in all the required places).

I guess this is not a critical PR for the immediate backport but it would be worth having it in 3.18 eventually...

michalvavrik commented:
@sberyozkin I am happy to read it tomorrow (as I read all your PRs anyway, add me as a reviewer or not) but I just finished long day on QE stuff, so not today, thanks for keeping me in a loop, good night :-)

gsmet reviewed (commented):
I'm going to ask a stupid question but do we need an additional property?

From what I can see, you would have the same result by increasing `expiresIn` a bit?

Interested in understanding why this is better!

sberyozkin commented:
Hi @gsmet 

It is a good question, thanks, that particular test was just easier for me to update to confirm adding the skew works - this test just checks that the calculated token expiry time is correct - in that particular setup it is not really necessary to configure 2 properties. There was a case reported by the Salesforce user where the standard `expires_in` property was not even returned, so the test which I updated supports that case - that if the provider does not tell Quarkus when the token expires, users can get a chance to tell Quarkus about it - so that the access token can be refreshed in a timely manner etc.

So usually, the main use case, the relative `expires_in` property is coming up in the token grant response. Now Quarkus would add it to the `now()` current time to calculate the absolute expires at time, but as was reported on the dev channel, there could be cases where using `now()` is not very precise. 

We have tests which work with this `expires_in` property is coming up in the token grant response, but I looked yesterday evening and there were quite complex already, actually waiting for the token to expire, so I just took an easy path with updating the test where adding skew is indeed not really necessary.

Hope it clarifies it

sberyozkin commented:
@gsmet, just for your info, #41067, the test which I updated was originally create to address that issue. You are right in this case the skew is unnecessary, but it is just the simplest test out there where I can check easily that configuring a skew has he impact on the test results :-)

I guess I'll go ahead and merge later today unless yourself or Michal will have more comments

michalvavrik reviewed (approved):
We also have `quarkus.oidc.token.refresh-token-time-skew` so it makes sense to align this for access token.

As for suggestion that users can use `access-token-expires-in`, there is a difference because source of `expiresAt` can be:

1. jwt claim expiration time
2. `absolute-expires-in` config property
3. granted tokens 'expires-in'

for `1.`, _now_ is not Quarkus server side _now_ but identity provider _now_ (when they created the tokens), `2.` is different completely. I saw a user scenario for this in the Zulip discussion.

@sberyozkin correct ^^ if I am wrong, but changes LGTM, thanks

michalvavrik commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientImpl.java:
```diff
@@ -287,6 +287,9 @@ private Long getAccessTokenExpiresAtValue(String token, Object expiresInValue) {
             final long now = System.currentTimeMillis() / 1000;
             expiresAt = now + oidcConfig.accessTokenExpiresIn().get().toSeconds();
         }
+        if (oidcConfig.accessTokenExpiresInSkew().isPresent()) {
```
but technically it is possible taht `expiresAt` is `null` here. why else would `if` condition handled such a scenario. I don't think you can do `+=`.

sberyozkin replied:
@michalvavrik Good catch (that would be that Salesforce case where the binary access token without any expiry hints was sent), let me just add a null check there, thanks

sberyozkin commented:
@michalvavrik Michal, I think I'll rename this property from `accessTokenExpiresInSkew` to`accessTokenExpirySkew` to make less tied to the expiry_in processing, thanks

sberyozkin commented:
Thanks for the feedback, @michalvavrik

sberyozkin commented:
Sorry for the noise with pushes, so, Michal, @michalvavrik,  I've renamed the property as suggested and simplified the JavaDocs definition to avoid sending a message one should only use it with the token grant response's `expires_in`, we don't really know if the JWT claim `exp` claim fallback will always be set to the correct value either... Added an NPE check as well...

Hope it is better now

michalvavrik reviewed (approved):
LGTM
